### ISSUE TITLE
[css-view-transitions-2] Define navigation descriptor for @view-transition

### ISSUE THREAD

**Issue by khushalsagar** [w3c/csswg-drafts#8783]
There are a class of navigations which are initiated via the browser UI:

* Entering a url in the URL bar.
* Back/forward buttons in the browser UI.

It's unclear whether we should initiate a ViewTransition for these. The worry is that we'd hit navigation paths the current Document doesn't expect. For example, there is no link to the Document the user entered in the URL bar in the contents of the current Document. So the author will need to ensure there is a transition defined between any 2 same-origin URLs. They can always use the cross-fade setup by the UA CSS.

If we don't enable ViewTransitions for navigations initiated using the URL bar, it makes handling back/forward slightly complicated. Using back/forward in the browser UI is a common navigation pattern. If A -> B had a ViewTransition, we likely also want one for B -> A triggered by the back button. So we'll have to carve out the subset of cases where back/forward in the UI use ViewTransitions.

**Comment by khushalsagar** [comment]
The resolution was to leave this undefined but we should add a non-normative note mentioning this.

**Comment by noamr** [comment]
I think this is a good time to revisit this (as mentioned in the minutes by @fantasai, "we should revisit this later").
Since we're defining the initial navigation types as `auto` and `none`, we should be explicit about what `auto` means.
Leaving this unspecified would leave an open interop question for this type of navigation.

I propose that `auto` would mean `normal | back | forward`, where `normal` means a normal click link, form submit etc), and `back` / `forward` mean a history traversal of `-1` or `+1` exactly.

We can then have `reload` and `jump` as optional navigation descriptor values that are not included in `auto` but can be explicitly enabled.

**Comment by khushalsagar** [comment]
Any reason to disallow transitions for script initiated traversals with delta > 1. It seems like by definition the author expects that route.

I think we can have `auto` include all navigations where [user navigation involvement](https://html.spec.whatwg.org/multipage/browsing-the-web.html#user-navigation-involvement) is not `browser UI`. For the `browser UI` case, only include `traverse` with a delta of 1. Since any other browser UI initiated navigation might be an unexpected route for the author.

@domenic @natechapin for an opinion too.

**Comment by domenic** [comment]
I was asked to comment here but don't quite understand the design goals. Some general principles and observations that might help:

- I think there are some implicit filters here that are not stated. Something like, cross-document, and same-origin? Can someone confirm?

- In terms of what authors expect, I think there's really only two meaningful tiers: link clicks/author-initiated JS navigations, and literally-everything. Saying that authors expect back button presses but don't expect manual same-origin URL hacking seems wrong to me. (In part, because once you've done the URL hacking to generate an arbitrary X -> Y push navigation, you can click the back button to generate an arbitrary Y -> X traverse navigation.)

- In the navigation API, we expose to authors the ability to be notified of [almost all navigations](https://github.com/wicg/navigation-api#appendix-types-of-navigations), except for browser-UI non-traverse. The reasoning there was not really about what authors expect, but instead about whether we could meaningfully allow cancelation or interception. For browser-UI non-traverse we do not offer cancelation or interception, because that wasn't possible before the navigation API existed and we didn't want to introduce a new capability.

  Even if the reasoning isn't the same, aligning on that same boundary might make things easier to understand for authors.

- In terms of what we see authors filtering out, we see lots of usage of these `NavigateEvent` properties: `hashChange`, `formData`, and `downloadRequest`. Authors mostly don't want to intercept those kinds of navigations. (Although intercepting `formData` is relatively popular.)

- It seems like you're trying to do two things: summarize the full power of the navigation API into a set of easy keywords for common use cases, and figure out the default auto behavior.

  For the former, I'd be cautious and minimal. Since you're not going to be able to give developers the full power of `e.destination.url.endsWith("/foo") && !e.hashChange && e.userInitiated && e.info === "bar"` in CSS, the default assumption should be that developers use JS to connect up their CSS appropriately. So it's only for truly-simple cases, where you have web developers explicitly saying "I really want to filter out the X transitions with no special cases", that you want to create new CSS keywords. `back`, `forward`, and `reload` are good guesses, but it'd be best if you could point to explicit developer demand, ideally in the form of building a realistic demo. In particular, to show that "I really want to filter out the back transitions" doesn't end up being "I really want to filter out the back transitions, except when going back to the homepage" or something like that.

  For the latter, whichever route you go, I really appreciate the efforts to phrase it in terms of the navigation API. Keep that up :)

**Comment by noamr** [comment]
> I was asked to comment here but don't quite understand the design goals. Some general principles and observations that might help:
> 
> * I think there are some implicit filters here that are not stated. Something like, cross-document, and same-origin? Can someone confirm?

Confirmed. This is in the current spec.

> * In terms of what authors expect, I think there's really only two meaningful tiers: link clicks/author-initiated JS navigations, and literally-everything. ... (In part, because once you've done the URL hacking to generate an arbitrary X -> Y push navigation, you can click the back button to generate an arbitrary Y -> X traverse navigation.)

That's a good point.

**Comment by khushalsagar** [comment]
Thanks for the feedback @domenic. Putting a CSS snippet for the resolved syntax for context on the replies below. This would be valid CSS for both the old and new Documents.

```css
/* This rule is used to decide whether a View Transition will be done for a navigation */
@view-transition {
   /* This is a transition opt-in for all navigations classified as "auto" */
   navigation: auto;
   
   /* This is a developer provided custom ident to conditionally apply CSS specific to a transition */
   type: home-to-back;
}

@media (prefers-reduced-motion) {
    @view-transition {
       /* This is a transition opt-out for all navigations */
       navigation: none;
    }
}

/* Example of how type is used to activate CSS specific to a transition */
html:active-view-transition(home-to-back) #target {
   view-transition-name: target;
}
```

> I think there are some implicit filters here that are not stated. Something like, cross-document, and same-origin? Can someone confirm?

Just to clarify why this isn't obvious. We broke down the parameters to identify a navigation as-follows:

- Source or destination URL
   This implicitly identifies whether the navigation is same-origin. Currently the CSS opt-in applies to all same-origin URLs. Authors will have to use script if they need to enable transitions to specific URLs. We'll eventually add from/to descriptors to the at-rule to make it declarative.

- Same vs cross document
   Currently the CSS opt-in only applies to same-document navigations. Authors can trigger a transition via `document.startViewTransition()`. We've discussed opening this up to same-document navigations but how is TBD. Should be a new descriptor if it needs to be in the at-rule.

- Everything else
   This is what the `navigation` descriptor is supposed to cover.

The goal of this issue it to define the values for the `navigation` descriptor and which navigation each one matches with. Based on your feedback, I think the following keywords makes sense:

* `all`: This matches all navigations that are observable with the `navigation` API and is an opt-in. Authors would use this if they are setting up the opt-in via script events (like `navigate`) on the old Document and `activationInfo` on the new Document.

* `none`:  This matches all navigations that are observable with the `navigation` API and is an opt-out.

* `auto`: This matches all navigations the author is guaranteed to expect: link clicks/author-initiated  (IIUC the spec way to say this would be user navigation involvement != "browser UI") *except* reload. We're excluding reload based on developer feedback (see https://github.com/w3c/csswg-drafts/issues/8784).

> So it's only for truly-simple cases ... that you want to create new CSS keywords

Agreed. Btw, since the CSS at-rule is both an opt-in and a setter for type, the latter is another motivation for new CSS keywords. For example, it would be common to customize a directional animation based on whether its a back or fwd navigation. But I agree that we should add it once there's been enough use of the JS API and we see what the common patterns are.

**Comment by khushalsagar** [comment]
> Does this intentionally drop history traversal?

It drops history traversal via browser UI but not if it was initiated by script based on this comment above:

> Saying that authors expect back button presses but don't expect manual same-origin URL hacking seems wrong to me. (In part, because once you've done the URL hacking to generate an arbitrary X -> Y push navigation, you can click the back button to generate an arbitrary Y -> X traverse navigation.)

Although now that you asked, we could have `auto` include back/fwd navigations from the browser UI *if* the push navigation was done by script?

**Comment by khushalsagar** [comment]
My suggestion was based on the idea that `all` is the keyword used when the actual decision is made in script. That's why it encompasses all navigations observable in script. And `auto` is the keyword used when the author wants the browser to enable the transition only if a route is "expected". So they don't need to add script to make that decision.

I see your point though, we haven't had enough developer feedback to be opinionated about what a generic keyword like `auto` should mean. The only feedback on unexpected transitions is around reload. So maybe we go with what you said and choose a better name than "auto" like "all-except-reload"?

The above implies that to begin with authors won't be able to trigger a transition on reload at all. We'll add that if there's demand for it.

**Comment by noamr** [comment]
Proposed resolution from internal sync, based on previous resolution:
Add a normative note saying that for navigations with browser UI user-involvement the UA MAY skip the navigation for user-experience reasons. Other than that, `auto` still means "everything but `reload`".

**Comment by noamr** [comment]
Proposal from internal sync:

- Use the name `auto` as per existing spec.
- The name `auto` should not match reloads, and also not match browser-UI non-traverse navigations (e.g. typing in the URL bar). This matches the behavior of the navigation API, where this type of navigation is not interceptable.
- The rationale for this is not about an unexpected route, but rather that:
  a. typing in the URL address bar and switching URLs. might "feel like" a new navigation rather than a transition between pages, so for now starting conservatively without those navigations feels safe.
  b. Because these navigations are not cancelable/interceptable, there's no way to customize/skip the transition in the JS, which is still an important piece of cross-document view transitions.

In summary, `auto`would match:
- Any `traverse` navigation
- Any `push` or `replace` navigation whose `user involvement` is not `browser UI`.

**Comment by noamr** [comment]
One thing that we'd want to resolve in with the WG is whether programmatic reloads (e.g. `location.reload()`) should be included in `auto`. As per the current spec they don't, but they're interceptable via the navigation API.

If we enable it, skipping the transition for a reload in the outgoing page is still possible, and would look like this:
```js
navigation.addEventListener("navigate", async e => {
  if (e.navigationType === "reload") {
     const sheet = new CSSStyleSheet();
     sheet.replaceSync("@view-transition { navigation: none }");
     // Note that there's no need to restore it because a `reload` cancels BFCache. 
     // If this was not a `reload` we'd have to remove the sheet from the adopted stylesheets on `pagehide`.
     document.adoptedStyleSheets.push(sheet);  
  }
});
```

**Comment by bokand** [comment]
Is there a reason why programmatic reloads should be treated specially? In particular, it seems a bit weird that if we later added a `reload` value for the descriptor it would only control browser-ui initiated reloads.

**Comment by noamr** [comment]
> Is there a reason why programmatic reloads should be treated specially? In particular, it seems a bit weird that if we later added a `reload` value for the descriptor it would only control browser-ui initiated reloads.

Mainly so that people who use `location.reload()` for something today and want to enable view-transitions would be able to do it. I'd say that the main reasoning against allowing view-transitions on reloads is for user-initiated reloads in the browser UI. The other reason would be to be consistent with existing web-platform features (namely the navigation API).
